satellite/satellitedb/apikeys.go

```diff
@@ -15,41 +16,130 @@ import (
 
 // apikeys is an implementation of satellite.APIKeys
 type apikeys struct {
-	db dbx.Methods
+	methods dbx.Methods
+	db      *dbx.DB
 }
 
-// GetByProjectID implements satellite.APIKeys ordered by name
-func (keys *apikeys) GetByProjectID(ctx context.Context, projectID uuid.UUID) (_ []console.APIKeyInfo, err error) {
+func (keys *apikeys) GetPagedByProjectID(ctx context.Context, projectID uuid.UUID, cursor console.APIKeyCursor) (akp *console.APIKeyPage, err error) {
 	defer mon.Task()(&ctx)(&err)
-	dbKeys, err := keys.db.All_ApiKey_By_ProjectId_OrderBy_Asc_Name(ctx, dbx.ApiKey_ProjectId(projectID[:]))
+
+	search := "%" + strings.Replace(cursor.Search, " ", "%", -1) + "%"
+
+	if cursor.Limit > 50 {
+		cursor.Limit = 50
+	}
+
+	if cursor.Page == 0 {
+		return nil, errs.New("page cannot be 0")
+	}
+
+	page := &console.APIKeyPage{
+		Search:         cursor.Search,
+		Limit:          cursor.Limit,
+		Offset:         uint64((cursor.Page - 1) * cursor.Limit),
+		Order:          cursor.Order,
+		OrderDirection: cursor.OrderDirection,
+	}
+
+	countQuery := keys.db.Rebind(`
+		SELECT COUNT(*)
+		FROM api_keys ak
+		WHERE ak.project_id = ?
+		AND ak.name LIKE ?
+	`)
+
+	countRow := keys.db.QueryRowContext(ctx,
+		countQuery,
+		projectID[:],
+		search)
+
+	err = countRow.Scan(&page.TotalCount)
+	if err != nil {
+		return nil, err
+	}
+	if page.TotalCount == 0 {
+		return page, nil
+	}
+	if page.Offset > page.TotalCount-1 {
+		return nil, errs.New("page is out of range")
+	}
+
+	repoundQuery := keys.db.Rebind(`
+		SELECT ak.id, ak.project_id, ak.name, ak.partner_id, ak.created_at 
+		FROM api_keys ak
+		WHERE ak.project_id = ?
+		AND ak.name LIKE ?
+		ORDER BY ` + sanitizedAPIKeyOrderColumnName(cursor.Order) + `
+		` + sanitizeOrderDirectionName(page.OrderDirection) + `
+		LIMIT ? OFFSET ?`)
+
+	rows, err := keys.db.QueryContext(ctx,
+		repoundQuery,
+		projectID[:],
+		search,
+		page.Limit,
+		page.Offset)
+
+	defer func() {
+		err = errs.Combine(err, rows.Close())
+	}()
+
 	if err != nil {
 		return nil, err
 	}
 
 	var apiKeys []console.APIKeyInfo
-	var parseErr errs.Group
+	for rows.Next() {
+		ak := console.APIKeyInfo{}
+		var keyIDBytes, projectIDBytes, partnerIDBytes []uint8
+		var keyID, projectID, partnerID uuid.UUID
+
+		err = rows.Scan(&keyIDBytes, &projectIDBytes, &ak.Name, &partnerIDBytes, &ak.CreatedAt)
```

Comment:
You can implement wrappers for types to handle the unmarshaling, simplifying this code.

Reply:
I looked through the project and didn't  find any wrapping, can you explain more what do mean by this?